src/CommitmentsV2/SFA.DAS.CommitmentsV2/Application/Commands/StopApprenticeship/StopApprenticeshipCommandHandler.cs

```diff
@@ -0,0 +1,110 @@
+        private readonly ICurrentDateTime _currentDate;
+        private readonly IAuthenticationService _authenticationService;
+        private readonly IEventPublisher _eventPublisher;
+        private readonly IMessageSession _nserviceBusContext;
+        private readonly IEncodingService _encodingService;
+        private readonly ILogger<StopApprenticeshipCommandHandler> _logger;
+        private const string StopNotificationEmailTemplate = "ProviderApprenticeshipStopNotification";
+
+        public StopApprenticeshipCommandHandler(
+            Lazy<ProviderCommitmentsDbContext> dbContext,
+            ICurrentDateTime currentDate,
+            IEventPublisher eventPublisher,
+            IAuthenticationService authenticationService,
+            IMessageSession nserviceBusContext,
+            IEncodingService encodingService,
+            ILogger<StopApprenticeshipCommandHandler> logger)
+        {
+            _dbContext = dbContext;
+            _currentDate = currentDate;
+            _eventPublisher = eventPublisher;
+            _authenticationService = authenticationService;
+            _nserviceBusContext = nserviceBusContext;
+            _encodingService = encodingService;
+            _logger = logger;
+        }
+
+        protected async override Task Handle(StopApprenticeshipCommand command, CancellationToken cancellationToken)
+        {
+            try
+            {
+                _logger.LogInformation($"Begin stopping apprenticeShip. Apprenticeship-Id:{command.ApprenticeshipId}");
+
+                var party = _authenticationService.GetUserParty();
+                CheckPartyIsValid(party);
+
+                var apprenticeship = await _dbContext.Value.Apprenticeships
+                    .Include(t => t.DataLockStatus)
+                    .Include(s => s.Cohort).ThenInclude(c => c.AccountLegalEntity)
+                    .SingleOrDefaultAsync(x => x.Id == command.ApprenticeshipId);
+
+                apprenticeship.ValidateApprenticeshipForStop(command.StopDate, command.AccountId, _currentDate);
+
+                apprenticeship.StopApprenticeship(command.StopDate, command.MadeRedundant, command.UserInfo, party);
+
+                _logger.LogInformation($"Stopped apprenticeShip. Apprenticeship-Id:{command.ApprenticeshipId}");
+
+                await _eventPublisher.Publish(new ApprenticeshipStoppedEvent
+                {
+                    AppliedOn = _currentDate.UtcNow,
+                    ApprenticeshipId = command.ApprenticeshipId,
+                    StopDate = command.StopDate
+                });
+
+                _logger.LogInformation($"Sending email to Provider {apprenticeship.Cohort.ProviderId}, template {StopNotificationEmailTemplate}");
+
+                await NotifyProvider(_nserviceBusContext, apprenticeship.Cohort.ProviderId, apprenticeship.Id, apprenticeship.Cohort.AccountLegalEntity.Name, apprenticeship.ApprenticeName, command.StopDate);
```

Comment:
The pattern we use elsewhere for side-effects like notifications is to add a new event handler - in this case, for the `ApprenticeshipStoppedEvent`. This handler's job would be to notify the provider (by sending the `SendEmailToProviderCommand`). This would be a good seperation of concerns, allowing the stop handler to just do what is necessary to invoke the stop, and be agnostic of side-effects.

I guess you'd need to be mindful of it reacting to events emitted by the v1 implementation though, until that is removed? Otherwise the provider would be double-notified.

Reply:
Just for information; Due to the duplication issue with V1 getting double notifications and not wanting to change the UI side as part of this work,  decision made to put this with the UI migration to V2 to refactor the email.